src/components/Switch.js

```diff
@@ -0,0 +1,54 @@
+import React, {Component} from 'react';
+import {TouchableOpacity, Animated} from 'react-native';
+import PropTypes from 'prop-types';
+import styles from '../styles/styles';
+
+const propTypes = {
+    isOn: PropTypes.bool.isRequired,
+    onToggle: PropTypes.func.isRequired,
+};
+
+class Switch extends Component {
+    constructor(props) {
+        super(props);
+        this.offPosition = 0;
+        this.onPosition = 20;
+        this.offsetX = new Animated.Value(props.isOn ? this.onPosition : this.offPosition);
+
+        this.toggleSwitch = this.toggleSwitch.bind(this);
+    }
+
+    componentDidUpdate(prevProps) {
+        if (prevProps.isOn !== this.props.isOn) {
+            this.toggleSwitch();
+        }
+    }
+
+    toggleSwitch() {
+        Animated.timing(this.offsetX, {
+            toValue: this.props.isOn ? this.onPosition : this.offPosition,
+            duration: 300,
+            useNativeDriver: true,
+        }).start();
+    }
+
+    render() {
+        const switchTransform = {transform: [{translateX: this.offsetX}]};
+        const {isOn, onToggle} = this.props;
+
+        return (
+            <TouchableOpacity
+                style={[styles.switchTrack, !isOn && styles.switchInactive]}
+                activeOpacity={0.8}
+                onPress={() => onToggle(!isOn)}
```

Comment:
Then just move the `toggleSwitch` method "inline" here

```suggestion
                onPress={() => {
                    onToggle(!isOn);
                    Animated.timing(this.offsetX, {
                        toValue: this.props.isOn ? this.onPosition : this.offPosition,
                        duration: 300,
                        useNativeDriver: true,
                    }).start();
                }}
```

Reply:
The only downside of this approach would be is that it's not reliant on the `isOn` value. for example if for whatever reason `onToggle` doesnt necessarily end up updating the `isOn` value, then the animation would be out of sync

Reply:
and since we're optimistically setting the value, this wouldnt revert the toggle on failed requests. let me know if you still want me to continue with the func comp refactor but I think this would cause issues @bondydaa